### PR TITLE
SI-7725 - Vector concatenation is unreasonably slow

### PR DESCRIPTION
**Pull request by Ichoran**
Rewrote ++ to use append or prepend when adding small collections to the end
or beginning of vectors.  This solves the extra-O(n) problem for addition of
single elements reported in SI_7725.

### CODE REVIEW

File: src/library/scala/collection/immutable/Vector.scala [hunk 1]
```diff
@@ -24,6 +24,10 @@ object Vector extends IndexedSeqFactory[Vector] {
     ReusableCBF.asInstanceOf[GenericCanBuildFrom[A]]
   private[immutable] val NIL = new Vector[Nothing](0, 0, 0)
   override def empty[A]: Vector[A] = NIL
+  
+  // Constants governing concat strategy for performance
+  private final val LgConcatFaster = 5
```
Comment (paulp) [inline src/library/scala/collection/immutable/Vector.scala]: Pedantic note as the door hits me on the way out: Lg? Is that "Large"? I feel strongly that if one has enough characters to write "ConcatFaster" one must have enough in reserve to write "Large".

Comment (paulp) [inline src/library/scala/collection/immutable/Vector.scala]: Alternatively, please spell Tiny "Ty" so the naming is uniform.

Comment (Ichoran) [inline src/library/scala/collection/immutable/Vector.scala]: Lg means Log base 2.

Comment (paulp) [inline src/library/scala/collection/immutable/Vector.scala]: That was my next guess. Is there some way to communicate that to someone who does math with pebbles?

Comment (paulp) [inline src/library/scala/collection/immutable/Vector.scala]: And I do remember we called natural log "ln" but I wouldn't expect "ln" to be understood as natural log in a programming context either. It's asking a lot for two letters to make the usage obvious.

Comment (som-snytt) [inline src/library/scala/collection/immutable/Vector.scala]: Paulp saved his best review comments for last.  Like the end of a fireworks display.

You could make it an inlined `logSizeTest`, but the day will come when someone will ask, where is the log file?

Comment (paulp) [inline src/library/scala/collection/immutable/Vector.scala]: Ha ha, nobody will ever ask where is the log file because nobody cares where the log file is.

Comment (Ichoran) [inline src/library/scala/collection/immutable/Vector.scala]: I guess I'll just use Log, assuming that people will realize that it is base 2 if it's an integer.

Comment (gkossakowski) [inline src/library/scala/collection/immutable/Vector.scala]: How about Log2?

**Comment by retronym** [comment]
LGTM.

**Comment by SethTisue** [comment]
whoops! https://issues.scala-lang.org/browse/SI-8014 , we all missed it

**Pull request merged**